Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,232 @@
+       * @param value The current value.
+       * @return Acceptable if the value was set successfully, Invalid
+       *         otherwise.
+       */
+      QValidator::State set_current(const Type& value) override;
+
+      /** Connects a slot to the current signal. */
+      boost::signals2::connection connect_current_signal(
+        const typename CurrentSignal::slot_type& slot) const override;
+
+    private:
+      mutable CurrentSignal m_current_signal;
+      Type m_current;
+      boost::optional<Type> m_default_value;
+      std::unordered_map<Type, std::shared_ptr<BooleanModel>> m_models;
+      std::vector<boost::signals2::scoped_connection> m_connections;
+      bool m_is_blocked;
+
+      void set_associated_model_value(const Type& value, bool model_value);
+      void on_current(const Type& value, bool is_selected);
+  };
+
+  template<typename T>
+  AssociativeValueModel<T>::AssociativeValueModel()
+    : m_is_blocked(false) {}
+
+  template<typename T>
+  AssociativeValueModel<T>::AssociativeValueModel(Type default_value)
+    : m_is_blocked(false),
+      m_default_value(std::move(default_value)) {}
+
+  template<typename T>
+  void AssociativeValueModel<T>::associate(
+      const std::shared_ptr<BooleanModel>& model, const Type& value) {
+    auto inserted_model = m_models.insert({value, model});
+    if(!inserted_model.second) {
+      return;
+    }
+    if(model->get_current()) {
+      model->set_current(false);
+    }
+    m_connections.push_back(model->connect_current_signal(
+      [=] (auto is_selected) { on_current(value, is_selected); }));
+    if(get_state() == QValidator::Invalid) {
+      set_current(value);
+    }
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::make_association(
+      const Type& value) {
+    auto model = std::make_shared<LocalBooleanModel>(false);
+    associate(model, value);
+    return model;
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::find(
+      const Type& value) const {
+    auto iterator = m_models.find(value);
+    if(iterator == m_models.end()) {
+      return nullptr;
+    }
+    return iterator->second;
+  }
+
+  template<typename T>
+  QValidator::State AssociativeValueModel<T>::get_state() const {
+    if(m_models.find(m_current) != m_models.end()) {
+      return QValidator::Acceptable;
+    }
+    return QValidator::Invalid;
+  }
+
+  template<typename T>
+  const typename AssociativeValueModel<T>::Type&
+      AssociativeValueModel<T>::get_current() const {
+    return m_current;
+  }
+
+  template<typename T>
+  QValidator::State AssociativeValueModel<T>::set_current(const Type& value) {
+    if(m_models.find(value) == m_models.end()) {
+      return QValidator::Invalid;
+    }
+    auto previous = m_current;
+    m_current = value;
+    set_associated_model_value(previous, false);
+    set_associated_model_value(m_current, true);
+    m_current_signal(m_current);
```

Comment:
This is also a re-entrancy bug.

    auto model = AssociativeValueModel<int>();
    auto updates = std::deque<int>();
    model.connect([&] (int value) {
      updates.push(value);
      if(value == 10) {
        model.set_current(20);
      }
    });
    model.connect([&] (int value) {
      updates.push(value);
    });
    model.make_association(10);
    model.make_association(20);
    model.set_current(10);
    REQUIRE(values.top() == 10);
    values.pop();
    REQUIRE(values.top() == 10);
    values.pop();
    REQUIRE(values.top() == 20);
    values.pop();